Refactored FIPS Checks into a Single Class (For Issue #34772)

LuisAlvelaMendes opened the pull request:
I have refactored all FIPS140 checks indicated in this issue: https://github.com/elastic/elasticsearch/issues/34772

Into a single FIPSChecks class, I also tried to implement a custom FIPSContext and interface that operate much like Bootstrap contexts, except without worrying about sometimes not being enforced (as FIPS is always enforced) and without requiring any metadata.

I also modified the unit tests to fit the changes and tried to call this class and the check method (that runs all checks) in the appropriate place, as discussed:

> A fitting place to call the checks is in createComponents of the Security plugin

Commit:
Refactoring FIPS checks

Commit:
check Bootstraps check in the commented thing

Commit:
refactored tests and also calling FIPSChecks on Bootstrap instead

Commit:
BUILD Sucessfull, still have to correct metadata and tests

Commit:
Fips Checks Refactored

Created a class FIPS Checks with a FIPS interface that allows you to perform a fips check per method, all checks are called at once,

Commit:
re-added deleted file

elasticmachine commented:
Pinging @elastic/es-security

jkakavas reviewed (changes requested):
Hi @LuisAlvelaMendes !

Thanks once again for taking this up :) I did a first pass of your PR and I've left several comments and suggestions about how I think this could be further refactored. Let me know what you think. 

Please also make sure that you have configured your git installation correctly to set your name in the commits ( https://help.github.com/articles/setting-your-username-in-git/ )

Finally also make sure that you read through or [contribution guide](https://github.com/elastic/elasticsearch/blob/master/CONTRIBUTING.md) and follow all the steps mentioned there. At the absolute minimum, ensure that [precommit check](https://github.com/elastic/elasticsearch/blob/master/TESTING.asciidoc#running-verification-tasks) is passing. 

Let me know if you have any questions

Commit:
Merge remote-tracking branch 'elastic/master'

Commit:
Refactoring FIPS checks


check Bootstraps check in the commented thing


refactored tests and also calling FIPSChecks on Bootstrap instead


BUILD Sucessfull, still have to correct metadata and tests


Fips Checks Refactored

Created a class FIPS Checks with a FIPS interface that allows you to perform a fips check per method, all checks are called at once,
re-added deleted file

Commit:
Merge 'elastic/master', squashed commit

Commit:
Merge remote-tracking branch 'origin/master'

LuisAlvelaMendes commented on .gitignore:
```diff
@@ -41,3 +41,4 @@ html_docs
 # random old stuff that we should look at the necessity of...
 /tmp/
 eclipse-build
+/bin/
```
This is my mistake, sorry for including certain files that weren't relevant

LuisAlvelaMendes commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSInterface.java:
```diff
@@ -0,0 +1,76 @@
+/**
+ * Encapsulates a FIPS check.
+ */
+public interface FIPSInterface {
```
The purpose would be to return an adequate FIPSCheckResult while also maintaining some consistency with how bootstrap checks were written .. I'm not quite sure what you mean b having the checks themselves implement it.

jkakavas replied:
I was thinking of an interface named `FIPSCheck` that then the checks like `FipsLicencingCheck` or `FipsPasswordHashingAlgoCheck` would implement. But I do think we don't necessarily need this either, there is no mandate to simulate bootstrap checks implementation. Same with `FIPSCheckResult`, it feels like an unnecessary abstraction. WDYT ?

LuisAlvelaMendes commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java:
```diff
@@ -0,0 +1,124 @@
+	
+    public FIPSCheckResult keystoreCheck(FIPSContext context) {
+
+        if (XPackSettings.FIPS_MODE_ENABLED.get(context.settings)) {
```
Do you mean add an if clause like `if (XPackSettings.FIPS_MODE_ENABLED.get(context.settings)) {}` surrounding it? I'm not sure what impact that would have

jkakavas replied:
I meant that we can have the if clause once in `Security.java` and call `check()` on all the checks (or the one `check()` if we decide to implement `FipsChecks` in a simpler manner ) IF `fips_mode` is set

LuisAlvelaMendes commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/FIPSChecks.java:
```diff
@@ -0,0 +1,124 @@
+			return FIPSCheckResult.success();
+		}
+		
+		else {
```
How do you suggest we do it then?

jkakavas replied:
I was thinking something similar to how we use [addValidationError](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/common/ValidationException.java)

LuisAlvelaMendes commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:
```diff
@@ -1063,10 +1066,12 @@ public void accept(DiscoveryNode node, ClusterState state) {
-                    FIPS140LicenseBootstrapCheck.ALLOWED_LICENSE_OPERATION_MODES.contains(license.operationMode()) == false) {
-                    throw new IllegalStateException("FIPS mode cannot be used with a [" + license.operationMode() +
+
+                License.OperationMode licenseOperationMode = FIPSChecksInstance.get().getLicenseOperationMode();
```
How, exactly? I thought we had established the license operations should be independant from any metadata.

LuisAlvelaMendes replied:
How, exactly? I thought we had established the license operations should be independant from any metadata.

LuisAlvelaMendes commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/Security.java:
```diff
@@ -499,6 +498,10 @@ protected Clock getClock() {
                 requestInterceptors, threadPool, securityContext.get(), destructiveOperations));
 
+        FIPSContext context = new FIPSContext(settings);
+        FIPSChecksInstance.set(new FIPSChecks(getLicenseState().getOperationMode()));
```
So we can get that instance with those values from the instruction `FIPSChecksInstance.get().check(context, env);` and call the checks

jkakavas replied:
The question was about why is it necessary to wrap the `FIPSChecks` in a `SetOnce` and not do 
```java
FIPSContext context = new FIPSContext(settings);
FIPSChecks checks = new FIPSChecks(getLicenseState().getOperationMode())
checks.check(context, env);
```

Commit:
removed some unecessary files, added license headers

jkakavas commented:
@LuisAlvelaMendes Thanks for the updates and apologies, I just noticed the new commits. I will take another look tomorrow, please feel free to ping me directly here every time there is a pending action on my side as otherwise your updates might get lost in the sea of Github notifications :)

jkakavas commented:
Hi @LuisAlvelaMendes 

I thought I'd ping you and see how this is going and if you're still interested in seeing this PR through. If there's anything you'd want to discuss, feel free to add a comment here. 

https://github.com/elastic/elasticsearch/pull/36677/commits/e124f8a012165e2e60e217fd22f475e3c9b4d576 made me think that you had addressed some of the points we discussed above, hence my comment in https://github.com/elastic/elasticsearch/pull/36677#issuecomment-449084241,  but I then saw that this is the same as https://github.com/elastic/elasticsearch/pull/36677/commits/7a4e6ff574396a47b9c7d1d8a9870d6029fdd8f7 that you had committed earlier.

LuisAlvelaMendes commented:
@jkakavas Sorry but due to some personal things I will not be able conclude this pull request.

jkakavas commented:
Closing this in favor of #47499
